runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClient.java

```diff
@@ -289,15 +312,18 @@ private synchronized void consumeWindmillWorkerEndpoints(WindmillEndpoints newWi
     ImmutableMap<Endpoint, WindmillConnection> newWindmillConnections =
         createNewWindmillConnections(newWindmillEndpoints.windmillEndpoints());
 
+    CompletableFuture<Void> closeStaleStreams = closeStaleStreams(newWindmillConnections.values());
+
     StreamingEngineConnectionState newConnectionsState =
         StreamingEngineConnectionState.builder()
             .setWindmillConnections(newWindmillConnections)
-            .setWindmillStreams(
-                closeStaleStreamsAndCreateNewStreams(newWindmillConnections.values()))
+            .setWindmillStreams(createAndStartNewStreams(newWindmillConnections.values()).join())
             .setGlobalDataStreams(
                 createNewGlobalDataStreams(newWindmillEndpoints.globalDataEndpoints()))
             .build();
 
+    closeStaleStreams.join();
```

Comment:
does building the connection state take any time? Or is there a reasno we can't move this join to after updating connections and the budget refresh below so we start using hte new streams earlier?

Otherwise it seems simpler to just keep the closeStaleStreamsAndCreateNewStreams which internally does them in parallel.

Reply:
good point moved